### ISSUE TITLE
Possible memory leak in websocket

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#1755]
After a couple days running as the main seed node on gaia-6002, my node crashed with OOM. It's a Digital Ocean box with 8GB memory. This happened twice.

It would appear it may have to do with websockets as there are numerous websocket connections and queries in the log. 

Probably related to https://github.com/tendermint/tendermint/issues/1740

**Comment by xla** [comment]
@ebuchman If the log is still around maybe worth attaching to #1740

**Comment by ebuchman** [comment]
See https://github.com/tendermint/tendermint/issues/1740#issuecomment-397754003

**Comment by melekes** [comment]
Indeed. Looks like we're leaking memory in `wsConnection#readRoutine` when the latter calls `rpc#subscribe`. But I can't find any leaking goroutines. Why this memory is not GCed?

![screen shot 2018-06-18 at 23 59 34](https://user-images.githubusercontent.com/1282182/41559019-d8531ffe-7353-11e8-92ff-d2af1617c945.png)

[profile003.svg.zip](https://github.com/tendermint/tendermint/files/2112803/profile003.svg.zip)

**Comment by ebuchman** [comment]
What is this profile from ? The QueryParser.Init() code is autogenerated and a little scary. But we put that in the Query object and pass it around where it ends up stored in a few maps in the pubsub, and presumably it should stay there until the connection ends. So I would think this issue only comes from having high WSConn turnover. Is that right ? Then maybe we just need to figure out why this stuff isn't being cleaned up when peers are removed ....

**Comment by melekes** [comment]
> Then maybe we just need to figure out why this stuff isn't being cleaned up when connections are removed ....

looks like it

**Comment by melekes** [comment]
```go
# rpc/client/rpc_test.go
func TestHeaderEvents(t *testing.T) {
	// memory heap before
	f, err := os.Create("/tmp/mem1.mprof")
	if err != nil {
		t.Fatal(err)
	}
	pprof.WriteHeapProfile(f)
	f.Close()

	c := getHTTPClient()
	err = c.Start()
	require.Nil(t, err)

	evtTyp := types.EventNewBlockHeader
	evt, err := client.WaitForOneEvent(c, evtTyp, waitForEventTimeout)
	require.Nil(t, err)
	_, ok := evt.(types.EventDataNewBlockHeader)
	require.True(t, ok)

	c.Stop()

	runtime.GC()

	// memory heap before
	f, err = os.Create("/tmp/mem2.mprof")
	if err != nil {
		t.Fatal(err)
	}
	pprof.WriteHeapProfile(f)
	f.Close()

	// dump all running goroutines
	time.Sleep(10 * time.Second)
	pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
}
```

**Comment by ebuchman** [comment]
@melekes lets close this for the fix in #1769 ?
